pkg/reconciler/pipelinerun/affinity_assistant_test.go

```diff
@@ -60,36 +60,108 @@ var testPipelineRun = &v1beta1.PipelineRun{
+				Spec: v1beta1.PipelineRunSpec{
+					Workspaces: []v1beta1.WorkspaceBinding{
+						{
+							Name:     "EmptyDir Workspace",
+							EmptyDir: &corev1.EmptyDirVolumeSource{},
+						}},
+				},
+			},
+		},
 	}
 
-	err := c.createOrUpdateAffinityAssistants(ctx, testPipelineRun.Spec.Workspaces, testPipelineRun, testPipelineRun.Namespace)
-	if err != nil {
-		t.Errorf("unexpected error from createOrUpdateAffinityAssistants: %v", err)
-	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
 
-	expectedAffinityAssistantName := getAffinityAssistantName(workspaceName, testPipelineRun.Name)
-	_, err = c.KubeClientSet.AppsV1().StatefulSets(testPipelineRun.Namespace).Get(ctx, expectedAffinityAssistantName, metav1.GetOptions{})
-	if err != nil {
-		t.Errorf("unexpected error when retrieving StatefulSet: %v", err)
-	}
+			c := Reconciler{
+				KubeClientSet: fakek8s.NewSimpleClientset(),
+				Images:        pipeline.Images{},
+			}
 
-	err = c.cleanupAffinityAssistants(ctx, testPipelineRun)
-	if err != nil {
-		t.Errorf("unexpected error from cleanupAffinityAssistants: %v", err)
-	}
+			err := c.createOrUpdateAffinityAssistants(ctx, tc.pr.Spec.Workspaces, tc.pr, tc.pr.Namespace)
+			if err != nil {
+				t.Errorf("unexpected error from createOrUpdateAffinityAssistants: %v", err)
+			}
+
+			expectedAAName := getAffinityAssistantName(tc.pr.Spec.Workspaces[0].Name, tc.pr.Name)
+			aa, err := c.KubeClientSet.AppsV1().StatefulSets(testPipelineRun.Namespace).Get(ctx, expectedAAName, metav1.GetOptions{})
+			if err != nil {
+				if tc.pr.Spec.Workspaces[0].PersistentVolumeClaim == nil && tc.pr.Spec.Workspaces[0].VolumeClaimTemplate == nil {
+					if !apierrors.IsNotFound(err) {
+						t.Errorf("unexpected error when retrieving StatefulSet: %v", err)
+					}
+				} else {
+					t.Errorf("unexpected error when retrieving StatefulSet: %v", err)
+				}
+			}
+
+			// validate PVs from Affinity Assistant
+			if tc.pr.Spec.Workspaces[0].VolumeClaimTemplate != nil {
+				if len(aa.Spec.VolumeClaimTemplates) != 1 {
+					t.Errorf("unexpected VolumeClaimTemplate count, expect 1 but got %v", len(aa.Spec.VolumeClaimTemplates))
+				}
+				if d := cmp.Diff(*getVolumeClaimTemplate(tc.pr.Spec.Workspaces[0], tc.pr), aa.Spec.VolumeClaimTemplates[0]); d != "" {
+					t.Errorf("VolumeClaimTemplates diff: %s", diff.PrintWantGot(d))
+				}
+			} else if tc.pr.Spec.Workspaces[0].PersistentVolumeClaim != nil {
+				if len(aa.Spec.Template.Spec.Volumes) != 1 {
+					t.Errorf("unexpected PersistentVolumeClaim count, expect 1 but got %v", len(aa.Spec.Template.Spec.Volumes))
+				}
+				if d := cmp.Diff(tc.pr.Spec.Workspaces[0].PersistentVolumeClaim, aa.Spec.Template.Spec.Volumes[0].VolumeSource.PersistentVolumeClaim); d != "" {
+					t.Errorf("PersistentVolumeClaim diff: %s", diff.PrintWantGot(d))
+				}
+			}
+
+			// clean up Affinity Assistant
```

Comment:
Is this meant to clean up the test, or to test cleanup of the affinity assistant? Is there any way to separate out tests for cleanup?

Reply:
It is part of the original test and It meant to test clean up logic of the affinity assistant. I think it is kinda reasonable to test out the cleanup logic here since we need to create an Affinity Assistant to test clean up anyways if we separate it out.

Reply:
We have an increasing code coverage with current tests so I think it is ok for this PR. It is a good idea to have separate tests for the clean up logic. I will submit another PR for the extra unit tests considering the size of this PR.